doc/developer/tracing.md

```diff
@@ -0,0 +1,231 @@
+_as a parent_. For example, in:
+```
+let s1 = tracing::debug_span!("my_span").entered();
+let s2 = tracing::debug_span!("my_span2");
+```
+the `s2` span as `s1` as a parent.
+
+### `[tracing::instrument(...)]`
+Previous examples may feel heavyweight to operate, but `tracing` provides the [`tracing::instrument`] attribute macro
+to make it easier. This macro is applied to functions (or the implementations of trait methods) (it works with `async`
+functions as well, and `#[async_trait]`), and automatically creates and enters a span each time a function is called.
+This makes it easy to make a span tree based on your call graph.
+
+An example:
+```
+#[tracing::instrument]
+fn my_sub_function(field_that_will_be_recorded: String) { ... }
+
+
+#[tracing::instrument(skip(big_argument))]
+fn my_function(big_argument: MyState) {
+  ...
+  my_sub_function(big_argument.thing);
+  ...
+  my_sub_function(big_argument.two);
+}
+```
+
+Both the spans for the `my_sub_function` calls will be childen of the `my_function` span!
+
+See [`tracing::instrument`] for more info.
+
+### Spans and `async` code
+**Please see [here] for more advice on how to use spans with async code**
+
+While `#[tracing::instrument]` works fine with `async` code. `Span::enter` can be problematic, explained in the link above.
+If `#[tracing::instrument]` is not an option, then import [`tracing::Instrument`] and call async functions like this:
+```
+let s = info_span!("important_span");
+let ret = thing.call_async_method().instrument(s).await;
+```
+
+### Logging of spans
+
+By default in materialize binaries, spans are NOT logged to stderr. However, their fields ARE inherited
+```
+let e = tracing::debug_span!("my_span", ?enum_value").entered();
+tracing::debug!("processed {} items", items.len());
+```
+produces:
+```
+2022-02-15T18:40:14.289898Z DEBUG my::module: processed 100 items _enum_value_=EnumVariant
+```
+
+## OpenTelemetry
+
+While events are clearly useful for logging, so far in this document, its unclear what spans offer. While
+there are [many crates] that do interesting things with tracing information, the primary way we use spans in
+materialize binaries is to collect [OpenTelemetry] data. In Materialize binaries, tracing spans can be
+_automatically_ collected with _no additional work required at callsites_
+
+### Setup
+
+- For local setup, use the instructions here:
+<https://github.com/MaterializeInc/materialize/tree/main/misc/opentelemetry> to setup a local OpenTelemetry collector
+and ui to view traces.
+- TODO(guswynn): cloud honeycomb setup when its available
+- TODO(guswynn): link to demo video
+
+### Span visualization
+
+OpenTelemetry UI's like the [Jaeger] one in the local setup _are extraordinarly powerful tools for debugging_. They allow you to visualize
+and inspect the exact control flow graph of your service, including the latency of each operation, and contextual recorded fields. See
```

Comment:
Nitpick, but it would be nice to have some screenshots of Jaeger/Honeycomb in the introduction section, as right now the main way spans are consumed by engineers is a bit hidden away and hard to understand if you are not already familiar with them.

Reply:
I can add a jaeger screenshot!

Reply:
gonna do that as a separate pr because I have to setup assets!